pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +96,458 @@ const (
+// rollout.
+func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	out := config.Out
+	oldRc := config.OldRc
+	scaleRetryParams := NewRetryParams(config.Interval, config.Timeout)
+
+	// Find an existing controller (for continuing an interrupted update) or
+	// create a new one if necessary.
+	sourceId := fmt.Sprintf("%s:%s", oldRc.Name, oldRc.UID)
+	newRc, existed, err := r.getOrCreateTargetController(config.NewRc, sourceId)
+	if err != nil {
+		return err
+	}
+	if existed {
+		fmt.Fprintf(out, "Continuing update with existing controller %s.\n", newRc.Name)
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+	// Extract the original replica count from the old controller, adding the
+	// annotation if it doesn't yet exist.
+	_, hasOriginalAnnotation := oldRc.Annotations[originalReplicasAnnotation]
+	if !hasOriginalAnnotation {
+		existing, err := r.c.ReplicationControllers(oldRc.Namespace).Get(oldRc.Name)
+		if err != nil {
+			return err
+		}
+		if existing.Annotations == nil {
+			existing.Annotations = map[string]string{}
+		}
+		existing.Annotations[originalReplicasAnnotation] = strconv.Itoa(existing.Spec.Replicas)
+		updated, err := r.c.ReplicationControllers(existing.Namespace).Update(existing)
+		if err != nil {
+			return err
+		}
+		oldRc = updated
+	}
+	original, err := strconv.Atoi(oldRc.Annotations[originalReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s\n",
+			oldRc.Name, originalReplicasAnnotation, oldRc.Annotations[originalReplicasAnnotation])
+	}
+
+	// The maximum pods which can go unavailable during the update.
+	maxUnavailable, err := extractMaxValue(config.MaxUnavailable, "maxUnavailable", original)
+	if err != nil {
+		return err
+	}
+	// The maximum scaling increment.
+	maxSurge, err := extractMaxValue(config.MaxSurge, "maxSurge", original)
+	if err != nil {
+		return err
+	}
+	// Further validation.
+	if maxUnavailable == 0 && maxSurge == 0 {
+		return fmt.Errorf("one of maxSurge or maxUnavailable must be specified")
+	}
+	// The minumum pods which must remain available througout the update
+	// calculated for internal convenience.
+	minAvailable := original - maxUnavailable
+	// Set the initial increment. This will drive the direction of the update;
+	// an initial increment of 0 means the first scale-up will be skipped,
+	// causing a down/up direction.
+	increment := 0
+	// A surge value means we need to move up/down to make availability for a
+	// scale down, and we'll burst higher than if only a maxUnavailable were
+	// set.
+	if maxSurge > 0 {
+		increment = maxUnavailable + maxSurge
```

Comment:
Still having some trouble with the maxUnavailable/maxSurge case. You said:

>maxUnavailable has no role to play while deciding how much should I scale up.

But I can't reconcile this with:

>Yes. In this case, we should downsize the old RC by 10% before scaling up the new RC to 30%. If we scale up the new RC by 30% directly, then we run the risk of surging by 30%, but the maxSurgePercentage is 20%.

I'm trying to think about scaleup/scaledown independently in such a way that each function can base all decisions on state observation. Scale down is no problem, and I think we've mostly condensed scale up into the following formula:

`increment = (original + maxSurge) - (oldRc.Spec.Replicas + newRc.Spec.Replicas)`

This works for when maxSurge=0/maxUnavailable>0 and maxSurge>0/maxUnavailable=0, but doesn't work for maxSurge>0/maxUnavailable>0. By simply observing this state:

oldrc.replicas=10
newrc.replicas=0
original=10
desired=10
maxSurge=20
maxUnavailable=10

During scale up, how do we know to no-op? I'm having trouble coming up with an algo that doesn't make use of maxUnavailable, and even then it doesn't quite cover it.

The expected events would be:
* scale down to 9
* scale up to 3
* scale down to 6
* scale up to 6
* scale down to 3
* scale up to 9
* scale down to 0
* scale up to 10